### ISSUE TITLE
Zero-Copy Memory Query should not return empty via best_affinity_to

### ISSUE THREAD

**Issue by Anjiang-Wei** [StanfordLegion/legion#1348]
The following code queries the best visible memory (in terms of affinity) with respect to a processor requiring the target memory kind.
This code is quite useful for GPU -- each GPU can see many `FB_MEM` in visible_memories, but it should choose the `FB_MEM` that has the best affinity. The counter-intuitive part is that we need an if branch for memory queries.

The reason why we need the branch is that `Z_COPY_MEM` has the "same" affinity to several GPUs. As a result, if we use `best_affinity_to` for Z_COPY_MEM, then the query result will always be empty.

```
Memory NSMapper::query_best_memory_for_proc(const Processor& proc, const Memory::Kind& mem_target_kind)
{
  Machine::MemoryQuery visible_memories(machine);
  visible_memories.same_address_space_as(proc)
                  .only_kind(mem_target_kind);
  if (mem_target_kind != Memory::Z_COPY_MEM)
  {
      visible_memories.best_affinity_to(proc); // for memories that are not Z_COPY_MEM, we can use "best_affinity"
  }
  else
  {
      visible_memories.has_affinity_to(proc); // Z_COPY_MEM must use "has_affinity";
  }
 // ...
```
Alex said that I should file this as a bug instead of accepting it as a feature because typically users won't write this `if` branch depending on which memory type they are querying for.

My personal suggestion is that an ideal fix would be: for `Z_COPY_MEM`'s query, users should also be able to get the desired non-empty query results via `best_affinity_to`. The crux here is that `Z_COPY_MEM` has the "same" affinity to all GPUs, but it can potentially also be the best for a specific GPU.

**Comment by lightsighter** [comment]
I agree with the suggestion that this produces "surprising" behavior and therefore is undesirable because all surprising behavior is generally not good for users. However, I'm not sure if just silently giving out one of the processors is the right answer either, because that is equally likely to produce "surprising" behavior to a different class of users. I think I'd actually be in favor of raising an error if the user performs a query using `best_affinity_to` and there isn't a unique answer to said query. This would be the "least surprising" behavior in my opinion and make sure that all users expectations are matching what the code is actually doing. It would be the user's responsibility then to use `best_affinity_to` with care on machine models where they know they can guarantee uniqueness of their query. What do you think @alexaiken and @streichler?

**Comment by elliottslaughter** [comment]
Let me see if I'm understanding. `mem_query.best_affinity_to(proc)` means the memory with the best affinity to `proc`. In this case, there might be (say) 4 GPUs and 1 zero-copy memory. All 4 GPUs have the same affinity to this memory. Therefore, if you're querying the *other* direction, you'd get 4 GPUs with equal affinity. But in this direct of the query, there is one unique memory that is "best", it just happens to be "best" to other processors in addition to `proc`.

Unless we're trying to maintain an invariant that if `best_affinity_to(P1)` *cannot* return a memory if that memory would also be returned for `best_affinity_to(P2)` (for some `P2 != P1`), then it seems consistent to me to return the memory here.

**Comment by alexaiken** [comment]
I'm OK with it giving an error message.  But then what is the alternative query that users should run in this case?
________________________________

**Comment by lightsighter** [comment]
> Therefore, if you're querying the other direction, you'd get 4 GPUs with equal affinity. But in this direction of the query, there is one unique memory that is "best", it just happens to be "best" to other processors in addition to proc.

Ok, yes, I was thinking about it the other direction. If there is exactly one unique zero-copy memory that is best for all the GPUs, then I agree that this query should work. The other direction was the one I was originally thinking of: what is the GPU processor with the best affinity to a particular zero-copy memory, and that direction is not well defined if there are multiple GPUs that share the same zero-copy memory.

> I'm OK with it giving an error message.  But then what is the alternative query that users should run in this case?

The `has_affinity_to` constraint will still allow queries that can return multiple results (see code sample above). It's then up to the user to iterate through all the processors/memories returned by the query and pick the one they want (or just call `first` on the query and take whatever one is first in the results if you don't care).

**Comment by elliottslaughter** [comment]
I think what's confusing about this API is that there are really two directions:

 1. Memory to processor: Select memory M with affinity to processor P such that M has no higher affinity to any other processor P2. This is the "best" affinity for memory M.
 2. Processor to memory: Select memory M such that processor P does not have any higher affinity to another memory M2. This is the "best" affinity to a memory for processor P.

I don't know what naming would disambiguate these, but I think half the issue is that the naming of `MemoryQuery::best_affinity_to(Proc)` implies we are in case 2 above, when we are really in case 1.

The issue of whether we should return ties or not is (I think) a separate issue.

I'm bikeshedding here, but if we had two APIs:

```
MemoryQuery::best_affinity_to(Proc)
MemoryQuery::best_affinity_from(Proc)
```

There might be confusion as to which is which, but at least it's obvious there are two options and they are in some sense opposed. Or we could do something more verbose, but I'm not sure what would actually improve the clarity of the naming.

**Comment by eddy16112** [comment]
I agree that the `bets_affinity_to` is confusing, and I am OK with the to/from APIs.

**Comment by lightsighter** [comment]
Why do you need `MemoryQuery::best_affinity_from(Proc)` when there is already `ProcessorQuery::best_affinity_to(Mem)`?

**Comment by lightsighter** [comment]
Note the graph of processors and memories is not directed. All queries are done either from the perspective of the memory or the processor based on the type of the query, e.g. `MemoryQuery` or `ProcessorQuery` respectively.

**Comment by elliottslaughter** [comment]
I want a memory, not a processor.
